src/pykeen/training/callbacks.py

```diff
@@ -374,9 +384,117 @@ def post_epoch(self, epoch: int, epoch_loss: float, **kwargs: Any) -> None:  # n
             self.last_best_epoch = epoch
 
 
-callback_resolver: ClassResolver[TrainingCallback] = ClassResolver.from_subclasses(
-    base=TrainingCallback,
-)
+class OptimizerTrainingCallback(TrainingCallback):
+    """Use optimizer to update parameters."""
+
+    def __init__(self, step: bool = True):
+        """Initialize the callback.
+
+        :param step:
+            whether to apply the parameter update step
+        """
+        super().__init__()
+        self.step = step
+
+    # docstr-coverage: inherited
+    def pre_batch(self, **kwargs: Any) -> None:  # noqa: D102
+        # Recall that torch *accumulates* gradients. Before passing in a
+        # new instance, you need to zero out the gradients from the old instance
+        self.optimizer.zero_grad(set_to_none=True)
+
+    # docstr-coverage: inherited
+    def post_batch(self, epoch: int, batch, **kwargs: Any) -> None:  # noqa: D102
+        # when called by batch_size_search(), the parameter update should not be applied.
+        if self.step:
+            # TODO: callback pre_step
```

Comment:
do we need to do this TODO before merging?

Reply:
Yes, otherwise gradient clipping will not work

Reply:
Conceptually, it would make most sense to have the clipping callbacks be `pre_step` callbacks of the `OptimizerTrainingCallback` (and not subclass from `TrainingCallback`).

Reply:
it was easy, cf. 28e9805